can/interfaces/ics_neovi/neovi_bus.py

```diff
@@ -0,0 +1,255 @@
+"""
+ICS NeoVi interface module.
+
+python-ics is a Python wrapper around the API provided by Intrepid Control 
+Systems for communicating with their NeoVI range of devices.
+
+Implementation references:
+* https://github.com/intrepidcs/python_ics
+"""
+
+import logging
+
+try:
+    import queue
+except ImportError:
+    import Queue as queue
+
+from can import Message
+from can.bus import BusABC
+
+logger = logging.getLogger(__name__)
+
+try:
+    import ics
+except ImportError:
+    logger.error(
+        "You won't be able to use the ICS NeoVi can backend without the "
+        "python-ics module installed!"
+    )
+    ics = None
+
+
+class NeoViBus(BusABC):
+    """
+    The CAN Bus implemented for the python_ics interface
+    https://github.com/intrepidcs/python_ics
+    """
+
+    def __init__(self, channel=None, can_filters=None, **config):
+        """
+
+        :param int channel:
+            The Channel id to create this bus with.
+        :param list can_filters:
+            A list of dictionaries each containing a "can_id" and a "can_mask".
+        :param use_system_timestamp:
+            Use system timestamp for can messages instead of the hardware time
+            stamp
+
+                >>> [{"can_id": 0x11, "can_mask": 0x21}]
+
+        """
+        super(NeoViBus, self).__init__(channel, can_filters, **config)
+        if ics is None:
+            raise Exception('Please install python-ics')
+
+        logger.info("CAN Filters: {}".format(can_filters))
+        logger.info("Got configuration of: {}".format(config))
+
+        self._use_system_timestamp = bool(
+            config.get('use_system_timestamp', False)
+        )
+
+        try:
+            channel = int(channel)
+        except ValueError:
+            raise ValueError('channel must be an integer')
+
+        type_filter = config.get('type_filter')
+        serial = config.get('serial')
+        self.dev = self._open_device(type_filter, serial)
+
+        self.channel_info = '%s %s CH:%s' % (
+            self.dev.Name,
+            self.dev.SerialNumber,
+            channel
+        )
+        logger.info("Using device: {}".format(self.channel_info))
+
+        ics.load_default_settings(self.dev)
+
+        self.sw_filters = None
+        self.set_filters(can_filters)
+        self.rx_buffer = queue.Queue()
```

Comment:
A Queue is overkill since the code is single threaded. You can use a list instead.

Reply:
Agree, this is some left over code from some tests. However, I am still not set on if a separated "messages pump" thread/task is needed or not. The driver can only hold 20000 messages before dropping older messages.

I may keep the queue and have a single use thread thread get the messages when `recv` is called.

Reply:
The other interfaces also have a limited receive buffer. It is up to the application to process messages continuously, often using a thread. So you don’t have to solve that here. The BufferedReader class can also be used.

Reply:
Make sense.

Will use `collections.deque` instead of a `list` for efficiency.
https://docs.python.org/3/tutorial/datastructures.html#using-lists-as-queues